walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,107 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.LinkedList;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 public final class FindMeetingQuery {
+  
+  //returns a list of possible times the meeting can be held- includes optional attendees if possible
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> attendees = new ArrayList<String>(request.getAttendees());
+    Collection<String> optionalAttendees = new ArrayList<String>(request.getOptionalAttendees());
+    long duration = request.getDuration();
+
+    // if duration is longer than the entire day, no meeting can occur
+    if (duration > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+    //no attendees or optional attendees
+    if (attendees.size() == 0 && optionalAttendees.size() == 0) {
+      return (Arrays.asList(TimeRange.WHOLE_DAY));
+    }
+    // only mandatory attendees
+    if (attendees.size() > 0 && optionalAttendees.size() ==0){
+      return findTimes(attendees, events, duration);
+    }
+    // only optional attendees
+    if (attendees.size() ==0 && optionalAttendees.size() > 0){
+      return findTimes(optionalAttendees, events, duration);
+    }
+    // both optional and mandatory attendees, try merging the two attendee lists together and finding meeting times.
+    if (attendees.size() > 0 && optionalAttendees.size() >0){
+      attendees.addAll(optionalAttendees);
+      Collection<TimeRange> toReturn = findTimes(attendees, events, duration);
+      if (toReturn.size() > 0) {
+        return toReturn;
+      }
+    }
+    // if no meeting time exists such that mandatory and optional attendees can all go, just find meeting times for mandatory attendees
+    attendees.removeAll(optionalAttendees);
+    return findTimes(attendees, events, duration);
+  }
+
+  //given a list of attendees and events, find meeting times that work
+  public Collection<TimeRange> findTimes(Collection<String> attendees, Collection<Event> events, long duration) {
+    List<TimeRange> conflicts = new ArrayList<TimeRange>();
+    Collection<TimeRange> toReturn = new ArrayList<TimeRange>();
+ 
+    for (Event event : events) {
+      if (containsAttendees(attendees, event.getAttendees())) {
+        conflicts.add(event.getWhen());
+      }
+    }
+
+    if (conflicts.size() >0) {
+      toReturn = conjoinTimes(conflicts, (int) duration);
+      return (toReturn);
+    }
+
+    return (Arrays.asList(TimeRange.WHOLE_DAY));
   }
+
+  //given a list of conflicts, invert the conflict times to find available times
+  public Collection<TimeRange> conjoinTimes(List<TimeRange> conflicts, int duration) {
```

Comment:
nit: perhaps conjoinTimes isn't the most descriptive name, maybe something like `findNonConflictingTimes`?

also I think this method can be `private` too